feat: Add the new chat gear encoding, gear sharing GUI, item record GUI

kristofbolyai opened the pull request:
This PR/branch is set-up to track the progress on implementing the new gear encoding system.

## Details
The new system is designed to be a successor to the old chat encoding, but overall, it brings many changes and improvements:
- Encoding is separated into two layers (item -> byte[], byte[] -> string) to allow encoding into any kind of formats, like UTF-16, base64, binary, etc. This also allows the mod to reuse this kind of encoding later on, and allows third-parties to support encoding/decoding of chat items.
- Improve the encoded string data density (1 UTF-16 character = 2 bytes). While the new encoded gear strings might not be shorter, they still have better data density compared to the old format.
- Support encoding/decoding crafted items (or any custom gear items), as well as support for sharing crafted items as "recipes" (in-game support coming later)
- Improve regular item encoding by introducing both a normal and an extended encoding. Normal encoding behaves like the old chat encoding, but except that it encoding gear identifications, and does not rely on order. The extended encoding also encodes base values, making the encoded item independent from API versions.
- Improve regular item encoding by sharing info from new game mechanics, like shiny item ids.

The implementation details can be found in https://github.com/Wynntils/Artemis/issues/2246.

## Progress
The milestones below should mostly refer to separate PRs, but some might be completed by a single one.
- [x] Create the byte encoding layer (byte[] -> string)
- [x] Create an universal data source/API for shiny stat ids (with id-key mapping) and use it
- [x] Create an universal data source/API for gear stat ids (with id-key mapping) and use it
- [x] Implement regular gear encoding/decoding
- [x] Implement charm/tome encoding/decoding
- [x] Implement crafted gear encoding/decoding
- [x] Implement crafted consumable encoding/decoding
- [x] Gear sharing GUI
- [x] Item Storage GUI

kristofbolyai committed:
Start some experimental modelling for the new gear encoding

kristofbolyai committed:
feat: Add universal support for encoding byte arrays (#2255)

EncodedByteBuffer is can be used for encoding and kind of byte[] data into formats like UTF-16 and base64. This is used as the base for the new gear chat encoding.

magicus commented on fabric/src/test/java/TestUnsignedByte.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Test
+    public void simpleByteArray_toUnsignedByteWorks() {
+        UnsignedByte[] bytes = UnsignedByte.of(new byte[] {(byte) 214, 121, 11, 49, 43, 75});
```
It's kind of only the first value that actually tests that the unsigned handling works... :)

What would be the expected outcome of sending in a negative valued byte? Should it be tested?

kristofbolyai replied:
214 is basically a negative byte though. It would be converted into it's unsigned representation. But negative bytes are not used by the encoding, only unsigned bytes 0-255.

magicus replied:
Yes, that is why I wrote that it is only 214 that is doing any real testing. I should clarify: I think you should add more "negative" bytes, and especially the borderline cases like 127, 128, 0, 1. 

And my second question was: what would be the result of UnsignedByte.of(new byte[] { -1, -2, -127, -128, -129 })? I think it would be a useful test case to add.

kristofbolyai replied:
Added some test cases for them :)

kristofbolyai committed:
feat: Use ShinyStatTypes instead of strings for shiny stats (#2257)

* chore: Move ShinyModel to stats package

This package should be a bit more fitting place for handling item-related stats.

* feat: Use ShinyStatTypes instead of strings for shiny stats

* fix: Bugfix AllShinyStatsFunction after refactor

kristofbolyai committed:
feat: Create a universal data source/API for gear stat ids (with id-key mapping) and use it (#2277)

kristofbolyai committed:
feat: Finish modelling the whole encoding/decoding process

kristofbolyai committed:
chore: add some more tests in TestUnsignedByte

kristofbolyai committed:
feat: Add everything needed to encode GearItems

kristofbolyai committed:
feat: Add everything needed to decode GearItems

kristofbolyai committed:
feat: Make ChatItemFeature support the new encoding

kristofbolyai committed:
feat: Separate ItemScreenshotFeature from item encoding, make ChatItemFeature have a separate keybind

kristofbolyai committed:
fix: fix calculating stat values

kristofbolyai committed:
feat: Add support for encoding/decoding charm and tome items

kristofbolyai committed:
feat: Make shiny items show up differently in chat

kristofbolyai committed:
chore: remove duplicate "level" field in CraftedGearItem

kristofbolyai committed:
refactor: remove static id fields from data transformers

kristofbolyai committed:
feat: Add everything needed to decode CraftedGearItems

kristofbolyai committed:
feat: Add EncodingSettings for item encoding

kristofbolyai committed:
feat: Add everything needed to decode CraftedConsumableItems

kristofbolyai committed:
chore: Add TestUnsignedByteUtils tests

kristofbolyai commented:
@magicus This PR should be ready for you to review, whenever you have the time for it. It's still in draft, but it is only missing screen related additions, everything else works now. I know it's huge to review, that's why I am asking you earlier, so you have time to go through all commits.

kristofbolyai committed:
feat: Separate ItemScreenshotFeature from item encoding, make ChatItemFeature have a separate keybind

ShadowCat117 committed:
feat: Item sharing GUI (#2304)

Co-authored-by: kristofbolyai <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
fix: Don't try to encode unsupported items

magicus committed:
Merge branch 'main' into gear-encoding

magicus commented on common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java:
```diff
@@ -0,0 +1,244 @@
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.item.ItemStack;
+
+public class ItemSharingScreen extends WynntilsScreen {
```
I don't get it. This looks like a merge error. Was this not supposed to be added by the item sharing PR?

ShadowCat117 replied:
I targetted this branch for the item sharing GUI since it uses the new encoding

magicus replied:
Aha, I see. I missed that part. Thanks.

magicus commented:
Thanks for the heads-up! I'll look through it as soon as I can. I will probably need several passes.

Just a first reflection after a quick skimming: there seems to be two kinds of changes here, adapations to the existing models (presumably to support new needed functionality), and actual changes related to the item encoding. Do you think it would make sense to separate the first part into a separate PR? Both for ease of reviewing, and for code history clarity. Then if something breaks, we will know that it was not the item encoding itself but the preparations, or the other way round.

magicus commented:
Or it might be that the other changes are just spurious artifacts. I don't understand the ItemSharingScreen, and there might be other similar problems.

kristofbolyai commented:
> Thanks for the heads-up! I'll look through it as soon as I can. I will probably need several passes.
> 
> 
> 
> Just a first reflection after a quick skimming: there seems to be two kinds of changes here, adapations to the existing models (presumably to support new needed functionality), and actual changes related to the item encoding. Do you think it would make sense to separate the first part into a separate PR? Both for ease of reviewing, and for code history clarity. Then if something breaks, we will know that it was not the item encoding itself but the preparations, or the other way round.

Every relevant major adaptation was already merged into main in separate PRs. What's here is mostly small, unspecific leftovers. 

Additionally, the PR was made with separated specifically commits, so each commit can be treated as a PR, as all of them work and compile. Obviously, I can't keep force pushing for review changes, but at least, the main bits should be separated.

kristofbolyai committed:
Merge branch 'main' into gear-encoding

magicus commented:
Ok, fine. I'll review per commit.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/PowderDataTransformer.java:
```diff
@@ -4,17 +4,76 @@
+            }
+        }
+
+        // Pad the data to the nearest byte with '1' bits
```
Why not pad with 0? Seems more natural...

magicus replied:
The comment above is now incorrect...

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/PowderDataTransformer.java:
```diff
@@ -4,17 +4,76 @@
+    }
+
+    private ErrorOr<UnsignedByte[]> encodePowderData(PowderData data) {
+        // Powders are encoded as bits, a powder needs 5 bits to encode
```
So if we ever get a way to find the tier of applied powders, we will solve this by retiring this transformer, and creating a new with another ID and using that instead? Is that the general thinking?

kristofbolyai replied:
No, it supports powder tier encoding. So we just assume every powder is tier 6. Although yes, the general thinking would be to up the version, if some unsolvable conflict happens.

magicus replied:
I think I'm missing where the powder tier is encoded then..? Can you point me to it?

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/ShinyDataTransformer.java:
```diff
@@ -7,14 +7,39 @@
+        UnsignedByte[] shinyStatValueBytes =
+                UnsignedByteUtils.encodeVariableSizedInteger(data.shinyStat().value());
+
+        UnsignedByte[] bytes = new UnsignedByte[shinyStatValueBytes.length + 2];
```
All the other non-trivial encodes had nice comments describing the format. I'd like to see that here to, it was very helpful.

magicus commented on common/src/main/java/com/wynntils/utils/UnsignedByteUtils.java:
```diff
@@ -0,0 +1,79 @@
+        // Strings are encoded by encoding the char's ASCII value
+        // and is terminated by a 0 byte
+        UnsignedByte[] bytes = new UnsignedByte[string.length() + 1];
+        byte[] asciiBytes = string.getBytes(StandardCharsets.US_ASCII);
```
What will happen if the string contains a character that is not representable in US ASCII? Do we get a runtime exception?

kristofbolyai replied:
I'll make it check and throw a readable exception. It'll be passed up to the transformer and nicely displayed to the user.

magicus commented on common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java:
```diff
@@ -109,16 +131,94 @@ public void onChatReceived(ChatMessageReceivedEvent e) {
-        Style style = Style.EMPTY
-                .applyFormat(ChatFormatting.UNDERLINE)
-                .withColor(gearItem.getItemInfo().tier().getChatFormatting());
+    private void decoodeChatEncoding(List<StyledTextPart> changes, StyledTextPart partToReplace) {
```
typo

magicus commented on common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java:
```diff
@@ -109,16 +131,94 @@ public void onChatReceived(ChatMessageReceivedEvent e) {
+        }
+    }
+
+    private void decoodeDeprecatedChatEncoding(List<StyledTextPart> changes, StyledTextPart partToReplace) {
```
typo

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/EndDataTransformer.java:
```diff
@@ -12,8 +12,6 @@
 import com.wynntils.utils.type.UnsignedByte;
 
 public class EndDataTransformer extends DataTransformer<EndData> {
-    public static final byte ID = (byte) 255;
```
I kind of like having a systematic constant like this, rather than a hard-coded value in a getter. I think you should basically revert this commit, but make the ID fields private (and keep the fix in the registry).

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/CustomGearTypeTransformer.java:
```diff
@@ -0,0 +1,67 @@
+
+public class CustomGearTypeTransformer extends DataTransformer<CustomGearTypeData> {
+    private static final List<Pair<GearType, Integer>> GEAR_TYPE_IDS = List.of(
+            new Pair<>(GearType.SPEAR, 0),
```
Maybe these ids should be stored in the GearType itself?

kristofbolyai replied:
I had a general rule when creating these transformers. If I added enums that are item encoding/transformer specific, I store the IDs there. However, if I use enums from other parts of the mod, like gear enums, I'll map them to an id in the transformer itself. This is mainly done to not have a lot of enum have some kind of "magic" ID. What do you think, should I add them to the enums, or keep it this way?

magicus replied:
I think we can add the "magic" IDs to the enum. Compare it with e.g. the API name. In a way that is a "magic" key for the Wynncraft API. And this ID is just another mapping between our data types and some other external interface. As long as we're not downing in such interfaces, I think it makes sense to store it as part of our definition.

kristofbolyai replied:
Okay, sure. It's done.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/CustomIdentificationDataTransformer.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public byte getId() {
+        return 12;
```
... and, to complement my previous comment, I'd like to see an ID field introduced here as well. (The "No magic numbers in code" rule).

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/DamageDataTransformer.java:
```diff
@@ -0,0 +1,128 @@
+            bytes.add(UnsignedByte.of((byte) unsignedBytes.length));
+            bytes.addAll(List.of(unsignedBytes));
+
+            // The next byte is the length of the integer that the maximum damage can fit in.
```
Just a reflection, maybe it is too late to change it now, and maybe it really doesn't matter, but there exists more compact representations for variable sized integers, that are assumed to be normally very low. I'm thinking of methods like UTF-8. Basically, the process is like: 
Store 7 bits of the source data in the target byte. Use the highest bit to mark if we're done. If it is 0, then we're done, and the byte is exactly the value we wanted. If it is 1, use the 7 bits we have, and grab the next 7 bits from the next byte. If that byte's highest bit is 0, then we're done (the value could be stored in 14 bits), otherwise continue.

kristofbolyai replied:
I've added it! It's actually great, since we encode a lot of numbers that fit 7 bits, so we save a fair amount of bytes. Implementation took a while, but I think it should be stable now :)

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/DefenseDataTransformer.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Override
+    public byte getId() {
+        return 11;
```
Another idea is to have a "global address enum", which contains the ID for all these data types. One upshot is that you avoid the "magic number in code" problem, and at the same time you get a overview of existing ids, and what the next free number would be. It also feels a bit contrary to the free-standing design currently used, but then again, a central rule coordinating numbers so they don't collide is necessary, so maybe it's better to actually respresent it as a enum class.

magicus replied:
In fact, the more I think about it, the better I like this idea. Don't restore the old ID fields, create a enum instead with the responsibility of owning the ID numbers.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/DefenseDataTransformer.java:
```diff
@@ -0,0 +1,96 @@
+    private ErrorOr<DefenseData> decodeDefenseData(ArrayReader<UnsignedByte> byteReader) {
+        // The first byte is the length of the integer that the health can fit in.
+        // The next bytes are the health bytes, which are assembled into an integer.
+        int length = byteReader.read().value();
```
This is a common pattern, reading a length byte + a variable sized integer. I recommend introducing a helper method to do just this. Then, if you decide to switch to another method of implementing variable-sized integers, it's just a single method to update.

(Basically, make it the responsibility of decodeVariableSizedInteger to read in the initial length byte also)

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/DurablityDataTransformer.java:
```diff
@@ -0,0 +1,82 @@
+        // The next byte is the length of the integer that the current durability can fit in.
+        // The next bytes are the current durability bytes, which are assembled into an integer.
+        int current = data.durability().current();
+        unsignedBytes = UnsignedByteUtils.encodeVariableSizedInteger(current);
```
And conversely, make a helper for writing variable length integers.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/RequirementsDataTransformer.java:
```diff
@@ -0,0 +1,128 @@
+
+public class RequirementsDataTransformer extends DataTransformer<RequirementsData> {
+    private static final List<Pair<ClassType, Integer>> CLASS_TYPE_IDS = List.of(
+            Pair.of(ClassType.NONE, 0),
```
Same question here, maybe put the id in ClassType?

magicus replied:
I'm not sure why this was resolved..?

magicus replied:
To clarify, I mean that ClassType should have an `encodingId` field containing these numbers.

kristofbolyai committed:
refactor: Introduce enums for data transformers

kristofbolyai committed:
refactor: PowderDataTransformer pads with 0 bits

kristofbolyai committed:
chore: add encoding standard comments to ShinyDataTransformer

kristofbolyai committed:
chore: check if we got an "ascii string" in UnsignedByteUtils

kristofbolyai committed:
feat: add error handling to ItemSharingScreen and extend it in ChatItemFeature

kristofbolyai committed:
feat: Use a more optimized variable length integer encoding

kristofbolyai commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/PowderDataTransformer.java:
```diff
@@ -0,0 +1,148 @@
+            for (int j = 0; j < 5; j++) {
+                // The index is reversed because the bits are stored in reverse order
+                int index = powderDataIndex + (4 - j);
+                boolean value = ((element * 6 + tier) >> j) % 2 != 0;
```
@magicus this line.

magicus replied:
Ah, thanks!

kristofbolyai committed:
chore: Run spotless to add 2024 to file copyright headers

This should make it so other commits don't have such huge diffs..

kristofbolyai committed:
refactor: Move encoding ids to enums

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/type/EncodingSettings.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Settings for encoding items.
+ * @param extendedIdentificationEncoding Whether to use extended identification encoding. This only affects identifiable items.
+ * @param shareItemName Whether to share the item name between the item name and the item lore. This only affects crafted and custom items.
```
"share the item name between the item name and the item lore". Sounds weird, not sure what is meant.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/CustomConsumableTypeDataTransformer.java:
```diff
@@ -0,0 +1,66 @@
+
+public class CustomConsumableTypeDataTransformer extends DataTransformer<CustomConsumableTypeData> {
+    private static final List<Pair<ConsumableType, Integer>> CONSUMABLE_TYPE_IDS = List.of(
+            new Pair<>(ConsumableType.POTION, 0),
```
Here's another one of the type/enum situations. Marking for reference; I'll get back to the general discussion on how to handle these separately.

kristofbolyai replied:
You are reviewing an older commit, this is fixed now.

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/EffectsDataTransformer.java:
```diff
@@ -0,0 +1,113 @@
+
+public class EffectsDataTransformer extends DataTransformer<EffectsData> {
+    private static final List<Pair<ConsumableEffect, Integer>> CONSUMABLE_EFFECT_IDS = List.of(
+            new Pair<>(ConsumableEffect.HEAL, 0),
```
... and here as well...

magicus commented on common/src/main/java/com/wynntils/models/items/ItemEncodingModel.java:
```diff
@@ -13,6 +15,12 @@
 import java.util.regex.Pattern;
 
 public class ItemEncodingModel extends Model {
+    @Persisted
```
This should have been configs, but we still cannot do configs in models, right? How do we change the value of these?

magicus replied:
Ah, I see, in the item sharing gui...

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/type/DataTransformerType.java:
```diff
@@ -0,0 +1,35 @@
+package com.wynntils.models.items.encoding.type;
+
+public enum DataTransformerType {
+    START_DATA_TRANSFORMER((byte) 0),
```
Maybe make the constructor take an int and do the casting in a single line only in the constructor?

kristofbolyai replied:
Sure.

magicus commented on common/src/main/java/com/wynntils/utils/UnsignedByteUtils.java:
```diff
@@ -69,71 +70,55 @@ public static String decodeString(List<UnsignedByte> bytes) {
-            bytes[5] = UnsignedByte.of((byte) (value >> 16));
-            bytes[6] = UnsignedByte.of((byte) (value >> 8));
-            bytes[7] = UnsignedByte.of((byte) value);
+        // Use zig-zag encoding to encode negative numbers
```
TIL!

kristofbolyai replied:
The amount of black magic used here makes me a bit nervous, even though the method is tested throughly with many cases for longs, ints, signs, et..

magicus commented on common/src/main/java/com/wynntils/models/gear/type/GearType.java:
```diff
@@ -38,34 +39,39 @@ public enum GearType {
     private final Item defaultItem;
     private final int defaultDamage;
     private final List<Item> otherItems;
+    private final int id;
```
Maybe name this and related methods etc `encodingId` to clarify?

kristofbolyai replied:
Only here, or every other enum?

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/DataTransformerRegistry.java:
```diff
@@ -0,0 +1,196 @@
+                bytes.addAll(Arrays.asList(errorOrEncodedData.getValue()));
+            } catch (Exception e) {
+                WynntilsMod.error(
+                        "Failed to encode data class " + itemData.getClass().getSimpleName() + "!", e);
```
A dedup is hiding here, waiting :)

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/DataTransformerRegistry.java:
```diff
@@ -0,0 +1,196 @@
+
+                dataList.add(errorOrData.getValue());
+            } catch (Exception e) {
+                WynntilsMod.error("Failed to decode data block with id " + dataBlockId.value() + "!", e);
```
... and here

magicus replied:
I see this is a common pattern. Maybe introduce a `ErrorOr.errorAndLog()`, `ErrorOr.loggedError()` or something like that?

kristofbolyai replied:
I think it's rather weird to call it like that. Maybe perhaps add a method like `errorInstance.log()`? Or `WynntilsMod#error(ErrorOr<>)`?

magicus replied:
I don't really know. I reacted to the duplication of the string. Either you can like create a `String errorMsg = "...."` and then call the log and ErrorOr with the string. Or you can create a common helper that basically does that. Whatever you want to call that, I don't really have any good suggestions.

kristofbolyai committed:
refactor: only cast once in DataTransformerType

kristofbolyai committed:
chore: fix comment in PowderDataTransformer

kristofbolyai committed:
chore: fix comment in EncodingSettings

magicus commented:
I just want to say that I have finished going through all individual commits (up until the last three new), but I also want to review the complete PR as well. I started doing that but was able to get only half-way through. I have been busy today and will be tomorrow as well; maybe I can get some time tomorrow night.

ShadowCat117 commented on common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java:
```diff
@@ -0,0 +1,260 @@
+
+    private void addSharingOptions() {
+        // region Checkbox options
+        if (wynnItem instanceof GearItem) {
```
This checkbox should probably be shown for charms and tomes too right? Or are they unlikely to change

kristofbolyai replied:
Yes. I'll patch it!

kristofbolyai commented:
> I just want to say that I have finished going through all individual commits (up until the last three new), but I also want to review the complete PR as well. I started doing that but was able to get only half-way through. I have been busy today and will be tomorrow as well; maybe I can get some time tomorrow night.

Sounds good. Plan is to merge this and the skill point loadout PR this weekend, bump to Artemis 1.0.0 and introduce a new versioning system. We may fit the item vault UI in this PR, if it is ready this week, we will see about  that.

`Artemis [bumps with mc version].[bumps with major features].[bumps with major fixes]-beta.[bumps with everything else]`.`

EDIT: Check out https://github.com/Wynntils/Artemis/pull/2313

This versioning scheme would work with the current state of Athena and Artemis, and allows us to easily add version streams without managing separate branches, as users can prefer to update on major/minor/all/etc versions.

kristofbolyai commented:
I also realized that I meant to reply this 2 days ago. In fact, I opened Github Mobile and this comment welcomed me, unsent. Sorry about it!

Anyways, I want to target Sunday for the release, if you want to have a final look.

magicus commented:
Making a final look now. Maybe you should take this out of draft at some point..? ;-)

kristofbolyai committed:
fix: show extended encoding option for tomes/charms

kristofbolyai committed:
fix: Don't encode unencodable custom gear types

kristofbolyai committed:
chore: rename GearType.id to encodingId

kristofbolyai committed:
refactor: Add ErrorOr#logged

magicus committed:
Merge branch 'main' into gear-encoding

magicus commented on common/src/main/java/com/wynntils/utils/type/ErrorOr.java:
```diff
@@ -41,6 +43,15 @@ public boolean hasError() {
         return error != null;
     }
 
+    public <T> ErrorOr<T> logged() {
```
This was a very elegant solution!

If we start to use this in other contexts, we can probably remove the throw clause and just do a no-op if there is no error.

magicus commented on common/src/main/java/com/wynntils/functions/InventoryFunctions.java:
```diff
@@ -256,7 +258,9 @@ public NamedValue getValue(FunctionArguments arguments) {
             if (shinyStatOpt.isEmpty()) return NamedValue.EMPTY;
 
-            return new NamedValue(shinyStatOpt.get().name(), shinyStatOpt.get().value());
+            // FIXME: The function system can't handle longs, so we have to cast to int
```
This irks me a bit. Are we planning to add longs to the function system? Can't we just remove the comment otherwise? Or are shinyStat values likely to overflow an int? If so, maybe we need to fix it. Otherwise, maybe it should return a int instead of a long.

Not a big thing, but it kind of bugs me.

kristofbolyai replied:
> Or are shinyStat values likely to overflow an int?

Could be the case. We have stats like "mobs killed". What if they add "experience collected"? These are sure candidates for overflow.

magicus replied:
Ok, let's leave the fixme for now then. Maybe open a bug to track that we need to handle this properly in the future.

magicus commented on common/src/main/java/com/wynntils/models/gear/type/ConsumableType.java:
```diff
@@ -1,13 +1,19 @@
+    FOOD(1),
+    SCROLL(2);
+
+    private final int id;
```
Should this be `encodingId` too?

magicus commented on common/src/main/java/com/wynntils/models/gear/type/GearType.java:
```diff
@@ -95,6 +101,13 @@ public static GearType fromClassType(ClassType classType) {
         return null;
     }
 
+    public static GearType fromId(int id) {
```
```suggestion
    public static GearType fromEncodingId(int encodingId) {
```

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/EndDataTransformer.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public byte getId() {
+        return (byte) 255;
```
Return a `DataTransformerType` value?

magicus commented on common/src/main/java/com/wynntils/utils/EncodedByteBuffer.java:
```diff
@@ -0,0 +1,135 @@
+            assert codePoint < 0xFFFFE : "Invalid code point: " + codePoint;
+        }
+
+        return fromBytes(bytes.toArray(new UnsignedByte[0]));
```
I think Java nowadays supports the more modern/logical:

```
        return fromBytes(bytes.toArray(UnsignedByte::new));
```

magicus commented on common/src/main/java/com/wynntils/models/items/encoding/impl/block/NameDataTransformer.java:
```diff
@@ -0,0 +1,62 @@
+
+    private ErrorOr<UnsignedByte[]> encodeName(String name) {
+        try {
+            return ErrorOr.of(UnsignedByteUtils.encodeString(name));
```
Wasn't the point that the name should not be byte-encoded, so at least it will show for non-Wynntils users? Or maybe I'm missing something in the grand scheme of things how the chat encoding packet is sent...

kristofbolyai replied:
It was the point yes. But it turned out to be really messy, so I opted to just encode the name. Hopefully it is not a big issue, and if it is, we can fix it up with a new version.

magicus replied:
Perhaps if we put the name as part of the start or end packet it gets simpler? I mean, if we want a non-encoded string it is more part of the packet itself, not the individual data blocks. So maybe we could sneak it in in the start block, and when we encode the start block, we extract the name and send it separately..?

magicus replied:
If we want to do something like that, it is probably best to do it now, since that changes the entire packet concept, rather than an individual data packet. Or, hm, maybe we should do both. That is, both keep this for the actual name, and then add some kind of "prefix" to the packet containing a "user representation", like `<start user representation byte>User visible name for non-wynntils users<the entire encoded packet>`.

kristofbolyai replied:
Let me share my thinking. Normally, I would say, yes, let's do this, give everyone the best experience. But this case, it's really tricky to get the code quality/system right, without breaking major abstractions. Vanilla users feel left out, yes. However, they are basically left out either way, Wynncraft items tend to be special not because of rarity, but their stats. So knowing that "I am selling x mythic" is not any better than knowing "I am selling x" without knowing it's stats.

The whole `EncodedByteBuffer` concept relies on the fact that everything is encoded. This (sadly) includes names too. We even "gain" a little, we can encode 2 ASCII chars into a single Unicode character.

If this becomes a problem for Wynncraft, I'll personally work on fixing it. However, I think we should keep it as is for now.

magicus reviewed (approved):
*phew* I have finally examined the complete PR in detail. Overall it looks good, nah, great! I added a few remarks, if you want to take care of them before pushing it would be nice, but they are not really blockers.

I am still a bit amazed at how the code exploded from like a single method in a feature to a huge framework. But then again, I cannot see how we could have done it differently either.

kristofbolyai committed:
chore: use ::new for arrays

kristofbolyai committed:
chore: Use DataTransformerTypes in EndDataTransformer

kristofbolyai committed:
chore: id to encodingId

ShadowCat117 committed:
feat: Item Storage GUI (#2312)

* feat: Item Storage GUI

* refactor: Move to ItemVaultService

* feat: Change save item text

* refactor: Move default category to service

* fix: Return super in mouseDragged

* fix: TreeMap not HashMap

* refactor: Change help button into button on vault

* fix: Don't crash when right clicking add category button

* chore: Move SavedItemStack to service package

* feat: Editing category names

* chore: Spotless

* feat: Change storage system, allow multiple categories and support multi selecting

* fix: Fix some scrolling issues

* fix: Fix deselecting crashing

* refactor: TreeSet not HashSet

* refactor: Use SavedItemsButton for help button

* feat: Add ItemStackInfo to SavedItem

* fix: Render tooltip above category name

* Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* refactor: custom json serialization for SavedItem

* chore: Use GLFW constants

* feat: Save to vault keybind

* feat: Add clear method to SavedItemsMenu

* refactor: Move more logic to service

* chore: Update tooltips

* feat: Change add/move category logic to shift held when clicking confirm button

* fix: Remove input when scrolling categories

* fix: Fix crash when deleting final item on row

* fix: Correct item selection when in multiple categories

* chore: Rename to Item Record

* refactor: More renames to item record, add drag to select

* feat: Add keybind to open item record

* fix: commit textures

* fix: More renames

---------

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>
